### PR TITLE
Fix reported PriorityQueue issue.

### LINKED ISSUE #77212: PriorityQueue<TElement, TPriority> edge case Count out of sync
### Description

I found an edge case in the [`PriorityQueue<TElement, TPriority>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.priorityqueue-2) collection, where enqueuing an element
might fail, but nonetheless the `Count` increases.

### Reproduction Steps

```C#
using System;
using System.Collections.Generic;

class Program
{
    static void Main()
    {
        PriorityQueue<int, int> queue = new(new (int, int)[300_000_000]);
        Console.WriteLine($"Before Enqueue, Count: {queue.Count:#,0}");
        try
        {
            queue.Enqueue(1, 1);
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.Message);
        }
        Console.WriteLine($"After Enqueue, Count: {queue.Count:#,0}");
    }
}
```

In my PC I get this output, after around 30 seconds of execution time:

```
Before Enqueue, Count: 300,000,000
Exception of type 'System.OutOfMemoryException' was thrown.
After Enqueue, Count: 300,000,001
```

### Expected behavior

After failing to `Enqueue`, the `Count` should preserve its previous value (300,000,000).

### Actual behavior

The `Count` increases to 300,000,001.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

- .NET 6.0
- Console application
- 64-bit operating system, x64-based processor

### Other information

_No response_

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fix #77212.

### CODE REVIEW

File: src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs [hunk 1]
```diff
@@ -464,10 +466,10 @@ public void EnqueueRange(IEnumerable<TElement> elements, TPriority priority)
             ArgumentNullException.ThrowIfNull(elements);
 
             int count;
-            if (elements is ICollection<(TElement Element, TPriority Priority)> collection &&
```
Comment (eiriktsarpalis) [inline src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs:467]: This was previously checking for the wrong ICollection type resulting in the optimization never being used.

File: src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs [hunk 2]
```diff
@@ -200,14 +200,16 @@ public void Enqueue(TElement element, TPriority priority)
...
-            int currentSize = _size++;
+            int currentSize = _size;
             _version++;
 
             if (_nodes.Length == currentSize)
             {
                 Grow(currentSize + 1);
             }
 
+            _size = currentSize + 1;
```
Comment (krwq) [inline src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs:211]: previously you passed in currentSize to MoveUpDefaultComparer/MoveUpCustomComparer and now this value is smaller by 1. I'm not seeing any logic adjustment - is that by design?

File: src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs [hunk 3]
```diff
@@ -200,14 +200,16 @@ public void Enqueue(TElement element, TPriority priority)
             // Note that the node being enqueued does not need to be physically placed
             // there at this point, as such an assignment would be redundant.
 
-            int currentSize = _size++;
+            int currentSize = _size;
```
Comment (krwq) [inline src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs:203]: could MoveUpDefaultComparer/MoveUpCustomComparer throw? If so should you be doing `_size + 1` here and `_size = currentSize` in the end of this method?

Comment (eiriktsarpalis) [reply src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs:203]: No, it cannot. I could have moved it to the end of the method, but the Move* implementations currently require the size being up to date and I didn't want to change that invariant.

**Review by krwq — APPROVED** [review]

**Pull request merged**